src/table.rs

```diff
@@ -252,14 +264,13 @@ impl<T: Slot> TablePage for Page<T> {
 
 impl<T: Slot> Drop for Page<T> {
     fn drop(&mut self) {
-        // Free `self.data` and the data within: to do this, we swap it out with an empty vector
-        // and then convert it from a `Vec<UnsafeCell<T>>` with partially uninitialized values
-        // to a `Vec<T>` with the correct length. This way the `Vec` drop impl can do its job.
-        let mut data = std::mem::take(&mut self.data);
-        let len = self.allocated.load();
+        // Execute destructors for all initialized elements
+        let len = self.allocated.load(Ordering::Acquire);
+        // SAFETY: self.data is initialized for T's up to len
         unsafe {
-            data.set_len(len);
-            drop(std::mem::transmute::<Vec<UnsafeCell<T>>, Vec<T>>(data));
+            // FIXME: Should be ptr::from_raw_parts_mut but that is unstable
+            let to_drop = slice::from_raw_parts_mut(self.data.as_mut_ptr().cast::<T>(), len);
```

Comment:
what's the consequence of this being `slice` instead or `ptr`?

Reply:
It creates an intermediate reference. It's not an issue in this case here but generally it ought to be avoided if not needed as references have more implications than raw pointers

Reply:
cool, merging!